### DISCUSSION TITLE
Why the EndPaint method has always required full painting.

### DISCUSSION THREAD

**Discussion by RyanCode** [FreeRDP/FreeRDP#9065] [Q&A] [answered]
Sorry to bother you, I have two questions. When specifying RGBA32 pixels, why is the gdi stride 4096 instead of 8192 when the width is 1024? The second question is, why does EndPaint always require full painting from (0,0) instead of incremental drawing?Looking forward to your reply.

<img width="991" alt="image" src="https://github.com/FreeRDP/FreeRDP/assets/61410262/7b4a5230-e83e-4445-81ed-ab2b5291ed7f">

**Comment by akallabeth** [comment]
@RyanCode
1.  `Bits` vs `Bytes` ;)
2. well, that depends on what the server sends (and what you do in your `nl_end_paint`, but you get a damage area for each end paint

**Reply by RyanCode** [reply]
when I set to enable gfx:
```c
    /*enable gfx*/
    settings->SupportGraphicsPipeline = TRUE;
    settings->RemoteFxCodec = TRUE;
    /*enable fast path*/
    settings->FastPathOutput = TRUE;
    settings->ColorDepth = 32;
    settings->SoftwareGdi = TRUE;
```
Img buffer from EndPaint is not complete.

<img width="865" alt="image" src="https://github.com/FreeRDP/FreeRDP/assets/61410262/5811f6be-1e20-4e4d-a804-83692897e378">

**Reply by akallabeth** [reply]
and if you set `TRUE = FALSE` everything is a lie ;)

1. why are you still changing random settings without a basic understanding of what happens? (hint: `NEVER` change `FastPathOutput`, better use the command line parser and let it handle the settings dependencies until you really know what you are doing)
2. What is your connection mode? (please tell me you know the modes are determined, you only announce a preference client side, the server decides what will be used)
3. How did you create this image? (screenshot, custom image dump routine, ...) 
4. did you compile this yourself? (most likely), are all dependencies there? (you might find an error in the log if some library dependency is missing, which might prevent proper image creation) (missing cairo or swscale are often causing image artifacts)
5. do you know that the server did already send a full image? `EndPaint` is called when the server tells us a bunch of drawing operations is completed, it is not necessarily a full screen update (as mentioned, check the damage area for details)

**Reply by RyanCode** [reply]
I haven't made to much changes to the code from sameple client, currently I'm just converting the primary buffer in the endpaint to webp format.just like this:webp x:0,y:0,w:1440,h:512,stride:5760.and data from endpaint is always like that.How could it be possible to start drawing the entire image from 0 every time?
```c
BOOL nl_end_paint(rdpContext *context)
{
	rdpGdi *gdi;
	int ninvalid;
	HGDI_DC hdc;
	nlClientContext *client_ctx;
	INT32 x, y, w, h;

	WINPR_ASSERT(context);
	WINPR_ASSERT(context->gdi);
	client_ctx = (nlClientContext *)context;
	gdi = context->gdi;
	if (!client_ctx->transport)
	{
		return FALSE;
	}

	/* Ignore paint if GDI output is suppressed */
	if (gdi->suppressOutput)
		return TRUE;

	/* Ignore paint if nothing has been done (empty rect) */
	if (gdi->primary->hdc->hwnd->invalid->null)
		return TRUE;

	hdc = gdi->primary->hdc;

	x = hdc->hwnd->invalid->x;
	y = hdc->hwnd->invalid->y;
	w = hdc->hwnd->invalid->w;
	h = hdc->hwnd->invalid->h;

	if (!webp_encoder_write(client_ctx->transport,
							(unsigned char *)(gdi->primary_buffer + 4 * x + y * gdi->stride),
							w, h, x, y, gdi->stride))
		nl_stop_client(context);
	return TRUE;
}
```

**Reply by akallabeth** [reply, marked as answer]
@RyanCode you really got a hang to pick the most buggy clients as template for you implementation ;)
Please check the sample client, `xfreerdp` or the newer `SDL` client. Ignore the others, there is a reason we print deprecation warnings with these all over the place.

as for how to always feed full updates:
1. `gdi->primary_buffer` is a pointer to the local framebuffer containing the current image
2. `gdi->stride` is the byte width of a line in the framebuffer image
3. `gdi->width` is the width of the framebuffer image
4. `gdi->height` is the height of the framebuffer image
5. `gdi->dstFormat` is the color format, use `FreeRDPGetBitsPerPixel(gdi->dstFormat)` to get color depth

as for `how to do partial updates:
```
const INT32 ninvalid = gdi->primary->hdc->hwnd->ninvalid;
const GDI_RGN* cinvalid = gdi->primary->hdc->hwnd->cinvalid;
for (INT32 x=0; x<ninvalid; x++) {
   const GDI_RGN* cur = &cinvalid[x];
   ...
}
```
